llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -438,6 +450,161 @@ bool duplicateAdressingRegs(MachineBasicBlock &MBB, MachineRegisterInfo &MRI) {
+/// following pattern:
+/// VADD (VSEL (VBCST(GlobalValue1), VBCST(GlobalValue2), Mask) Value)
+bool findBroadcastedGlobalValues(const MachineInstr *MI,
+                                 SmallPtrSet<const Value *, 4> &GVSet,
+                                 MachineRegisterInfo &MRI) {
+
+  const TargetSubtargetInfo &ST = MI->getParent()->getParent()->getSubtarget();
+  const AIEBaseInstrInfo *TII =
+      static_cast<const AIEBaseInstrInfo *>(ST.getInstrInfo());
+
+  // Skip subvector copy.
+  if (MI->isCopy()) {
+    Register Src = MI->getOperand(1).getReg();
+    if (Src.isVirtual())
+      MI = MRI.getVRegDef(Src);
+    else
+      // We cannot proceed with physical registers.
+      return false;
+  }
+
+  auto ParseByReg = [&](Register Reg, Opcodes Opcode) -> OptionalVecOp {
+    OptionalVecOp VOp = TII->parseTargetVectorOp(*MRI.getVRegDef(Reg));
+    if (VOp && VOp->Opcode == Opcode)
+      return VOp;
+    return {};
+  };
+
+  const OptionalVecOp VAddOp = TII->parseTargetVectorOp(*MI);
+
+  // We start with an ADD.
+  if (!VAddOp || VAddOp->Opcode != Opcodes::ADD)
+    return false;
+
+  // We should have one select in one operand.
+  AbstractVecOp VSelect;
+  if (OptionalVecOp Op = ParseByReg(VAddOp->VectorSrc1, Opcodes::SELECT))
+    VSelect = *Op;
+  else if (OptionalVecOp Op = ParseByReg(VAddOp->VectorSrc2, Opcodes::SELECT))
+    VSelect = *Op;
+  else
+    return false;
+
+  // We should have two broadcasts.
+  const OptionalVecOp VBcast1 =
+      ParseByReg(VSelect.VectorSrc1, Opcodes::BROADCAST);
+  const OptionalVecOp VBcast2 =
+      ParseByReg(VSelect.VectorSrc2, Opcodes::BROADCAST);
+
+  if (!VBcast1 || !VBcast2)
+    return false;
+
+  // We should end the search in just one last step each broacast.
+  SmallPtrSet<const MachineInstr *, 4> VisitedInstrs;
+  return findGlobalValues(MRI.getVRegDef(VBcast1->ScalarSrc), GVSet,
+                          VisitedInstrs, MRI, 1) &&
+         findGlobalValues(MRI.getVRegDef(VBcast2->ScalarSrc), GVSet,
+                          VisitedInstrs, MRI, VisitedInstrs.size() + 1);
+}
+
+bool fixLoadMemOpInfo(MachineFunction &MF, MachineBasicBlock &MBB,
+                      MachineRegisterInfo &MRI) {
+
+  bool Changed = false;
+
+  for (MachineInstr &MI : MBB) {
+    if (!MI.mayLoad() || !MI.memoperands_empty())
+      continue;
+
+    bool Success = true;
+    SmallPtrSet<const Value *, 4> GVSet;
+    SmallPtrSet<const MachineInstr *, 4> VisitedInstrs;
+
+    // As we don't know which operand is the pointer, we need
+    // to iterate over all uses.
+    for (auto MO : MI.uses()) {
+      if (MO.isReg() && MO.getReg().isVirtual()) {
+        const Register Reg = MO.getReg();
+        if (const MachineInstr *RegDef = MRI.getVRegDef(Reg)) {
+          if (LoadMMODeepSearch)
+            // Look to everything!
+            Success &= findGlobalValues(RegDef, GVSet, VisitedInstrs, MRI,
+                                        LoadMMOSearchLimit);
```

Comment:
I think we should avoid committing the `findGlobalValues` as it is overly optimistic. Maybe keep it on a separate branch for now?

Reply:
I can't say that this is optimistic or not, I can say that this is harder to break if we consider the pattern recognition. It tracks everything (in terms of globals) on the way of VLDB.4X load, it only rejects when it reaches function parameters (because they can be pointers as well). For AA perspective, if we miss one MMO is problem. On the other hand, if you include non-used ones, there is no problem because it will lead to more dependencies (just performance impacts).

Reply:
My main "issue" with it is that it considers loads safe, while I think they are only safe when we have already found a "vector" of pointers. In that case, it does not matter what we add to the vector, it could be a constant vector, or some loaded quantity.

Reply:
The new version addresses this situation: we need to find a `select(bcast(C1), bcast(C1), mask)` somewhere in the chain up starting form the load.